### PR TITLE
Allow to run any available backend with DBSCAN benchmark

### PR DESCRIPTION
**Pull request by aprokop**

### CODE REVIEW

File: benchmarks/dbscan/dbscan.cpp [hunk 1]
```diff
@@ -77,7 +78,7 @@ int main(int argc, char *argv[])
       ( "samples", bpo::value<int>(&params.num_samples)->default_value(-1), "number of samples" )
       ( "labels", bpo::value<std::string>(&params.filename_labels)->default_value(""), "clutering results output" )
       ( "print-dbscan-timers", bpo::bool_switch(&params.print_dbscan_timers)->default_value(false), "print dbscan timers")
-      ( "impl", bpo::value<std::string>(&params.implementation)->default_value("fdbscan"), R"(implementation ("fdbscan" or "fdbscan-densebox"))")
+      ( "impl", bpo::value<std::string>(&params.implementation)->default_value("fdbscan"), R"(implementation (fdbscan | fdbscan-densebox))")
```
Comment (aprokop) [inline benchmarks/dbscan/dbscan.cpp:81]: Drive-by change.

File: benchmarks/dbscan/dbscan.cpp [hunk 2]
```diff
@@ -68,6 +68,7 @@ int main(int argc, char *argv[])
       ( "help", "help message" )
       ( "algorithm", bpo::value<std::string>(&params.algorithm)->default_value("dbscan"), "algorithm (dbscan | mst)" )
       ( "filename", bpo::value<std::string>(&params.filename), "filename containing data" )
+      ( "backend", bpo::value<std::string>(&params.backend)->default_value("default"), "Kokkos backend (serial | openmp | cuda | hip | sycl | threads | openmptarget)" )
```
Comment (masterleinad) [inline benchmarks/dbscan/dbscan.cpp:71]: I would add the default value `default` to the list of backends to choose from.
```suggestion
      ( "backend", bpo::value<std::string>(&params.backend)->default_value("default"), "Kokkos backend (default | serial | openmp | cuda | hip | sycl | threads | openmptarget)" )
```
Also, generating the list based on the enabled backends would be nice.

File: benchmarks/dbscan/dbscan.cpp [hunk 3]
```diff
@@ -90,6 +91,13 @@ int main(int argc, char *argv[])
     return 1;
   }
 
+  if (std::set<std::string>{"fdbscan", "fdbscan-densebox"}.count(
+          params.implementation) == 0)
+  {
+    std::cerr << "Backend must be \"fdbscan\" or \"fdbscan-densebox\"\n";
+    return 2;
+  }
```
Comment (masterleinad) [inline benchmarks/dbscan/dbscan.cpp:99]: This looks like a duplicate of the lines below and you are only changing "Implementation" to "Backend"?!

File: benchmarks/dbscan/dbscan_timpl.hpp [hunk 4]
```diff
@@ -360,3 +358,74 @@ bool ArborXBenchmark::run(ArborXBenchmark::Parameters const &params)
...
+    return runImpl<DIM>(Kokkos::OpenMPTarget{}, params);
+#else
+    throw std::runtime_error("OpenMPTarget backend not available!");
+#endif
+  }
+
+  if (params.backend == "sycl")
+  {
+#ifdef KOKKOS_ENABLE_SYCL
+    return runImpl<DIM>(Kokkos::SYCL{}, params);
```
Comment (masterleinad) [inline benchmarks/dbscan/dbscan_timpl.hpp:425]:
```suggestion
    return runImpl<DIM>(Kokkos::Experimental::SYCL{}, params);
```

**Comment by aprokop** [comment]
Tests pass, per usual warning from #750.

File: benchmarks/dbscan/dbscan_timpl.hpp [hunk 5]
```diff
@@ -360,3 +358,74 @@ bool ArborXBenchmark::run(ArborXBenchmark::Parameters const &params)
...
+    return runImpl<DIM>(Kokkos::Experimental::HIP{}, params);
+#else
+    throw std::runtime_error("HIP backend not available!");
+#endif
+  }
+
+  if (params.backend == "openmptarget")
+  {
+#ifdef KOKKOS_ENABLE_OPENMPTARGET
+    return runImpl<DIM>(Kokkos::OpenMPTarget{}, params);
```
Comment (masterleinad) [inline benchmarks/dbscan/dbscan_timpl.hpp:416]:
```suggestion
    return runImpl<DIM>(Kokkos::Experimental::OpenMPTarget{}, params);
```

**Review by masterleinad — COMMENTED** [review]
O

File: benchmarks/dbscan/dbscan.cpp [hunk 6]
```diff
@@ -50,6 +50,41 @@ int getDataDimension(std::string const &filename, bool binary)
   return dim;
 }
 
+std::string backendsList(std::string const &delim = "|")
```
Comment (masterleinad) [inline benchmarks/dbscan/dbscan.cpp:53]: Why do we need a defaulted argument here?
```suggestion
std::string backendsList()
```

Comment (aprokop) [reply benchmarks/dbscan/dbscan.cpp:53]: Don't really need it. Just thought that if we decided it a general function to be used in other place, may as well allow some customization.

**Comment by dalg24** [comment]
> Allow to run any available backend with DBSCAN benchmark

Please elaborate why we should do that in the first place

**Comment by aprokop** [comment]
> Please elaborate why we should do that in the first place

The main reason is that this driver is used for the research purposed in addition to the benchmarking. This means that we want to shorten the time from "create/improve new feature" -> test it out. Meaning that we can build a single version of Kokkos with all backends enabled on a specific platform (say, Crusher), and we can build ArborX driver once.

In addition to speeding the build process once, it also removes the complexity of managing multiple executables and making sure that they correspond to the same code version (particularly important when the tree is dirty, as hash would not be enough to guarantee that).

**Comment by dalg24** [comment]
It also means longer build-time in all other situations and more complex code.

**Comment by dalg24** [comment]
And also is a bit problematic with the Serial backends with respect to atomics

**Comment by dalg24** [comment]
I suppose my question is considering all of it, are we still thinking this is worth it and why is it so

**Comment by aprokop** [comment]
> It also means longer build-time in all other situations and more complex code.

I understand your point. The only way to get the best of both worlds is to introduce a CMake switch to enable running multiple backends in a benchmark. 

> And also is a bit problematic with the Serial backends with respect to atomics

If you mean that atomics in serial are enabled when running with multiple backends, I'm not sure this is a problem that ArborX has to fix.

> I suppose my question is considering all of it, are we still thinking this is worth it and why is it so

It was absolutely worth it when running multiple experiments for a paper. I would still like to have this feature, even if not enabled by default. We do something similar for the `bvh_driver` benchmark, and if we ever decide to run dbscan benchmark for similar purposes (like, PR acceptance) it would have to be there.
